File: java/example/src/test/java/com/example/PlaybackConsumerTest.java
package com.example;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.paulhammant.servirtium.vcr.Outcome;
import com.paulhammant.servirtium.vcr.Vcr;
import com.paulhammant.servirtium.vcr.VcrServer;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Path;
import java.nio.file.Paths;
import org.junit.jupiter.api.Test;

/**
 * Third-party consumer test: exercises the INSTALLED servirtium-vcr jar (from
 * ~/.m2, resolved as a normal Maven dependency), replaying the canonical tape.
 * The engine .so is discovered zero-config from the jar's own /native/&lt;rid&gt;/
 * resource — no SERVIRTIUM_VCR_LIB, no source tree. This is the idiomatic Java
 * consumer experience (a jar bundles the native lib as a resource; there is no
 * loose .so file to pass to nativeLib(), so discovery is the mode here).
 */
class PlaybackConsumerTest {

    @Test
    void replaysCanonicalTapeFromInstalledJar() throws Exception {
        // Prove we're running the INSTALLED jar, not the source tree's classes.
        String codeLocation =
                Vcr.class.getProtectionDomain().getCodeSource().getLocation().toString();
        assertTrue(
                codeLocation.endsWith(".jar"),
                "expected servirtium-vcr to load from an installed jar, got " + codeLocation);

        Path tape = Paths.get(
                PlaybackConsumerTest.class.getResource("/tapes/single_get.md").toURI());

        try (VcrServer vcr = Vcr.playback(tape.toString()).port(0).start()) {
            HttpResponse<String> resp = HttpClient.newHttpClient().send(
                    HttpRequest.newBuilder(URI.create(vcr.baseUrl() + "/ok")).GET().build(),
                    HttpResponse.BodyHandlers.ofString());

            assertEquals(200, resp.statusCode());
            assertEquals("ok-body", resp.body());
            assertEquals(Outcome.OK, vcr.lastKind());
        }
    }
}

File: dotnet/example/PlaybackConsumerTest.cs
using System;
using System.IO;
using System.Net.Http;
using Servirtium.Vcr;
using Xunit;

namespace ConsumerExample;

/// <summary>
/// Third-party consumer test: exercises the INSTALLED Servirtium.Vcr NuGet
/// package (restored from a local feed as a normal PackageReference), replaying
/// the canonical tape. The engine .so is discovered zero-config from the
/// package's own runtimes/linux-x64/native/ asset — no SERVIRTIUM_VCR_LIB, no
/// source tree.
/// </summary>
public class PlaybackConsumerTest
{
    [Fact]
    public void ReplaysCanonicalTapeFromInstalledNuGetPackage()
    {
        // Prove the package's native asset deployed alongside the app (the
        // NuGet runtimes/<rid>/native payload), i.e. it came from the package.
        string nativeUnderApp = Path.Combine(
            AppContext.BaseDirectory, "runtimes", "linux-x64", "native", "libservirtium_vcr.so");
        string nativeNextToApp = Path.Combine(AppContext.BaseDirectory, "libservirtium_vcr.so");
        Assert.True(
            File.Exists(nativeUnderApp) || File.Exists(nativeNextToApp),
            "expected the NuGet package's bundled engine .so to be deployed under " + AppContext.BaseDirectory);

        string tape = Path.Combine(AppContext.BaseDirectory, "tapes", "single_get.md");

        using var vcr = Vcr.Playback(tape).Port(0).Start();
        using var client = new HttpClient { BaseAddress = new Uri(vcr.BaseUrl) };

        HttpResponseMessage resp = client.GetAsync("/ok").GetAwaiter().GetResult();
        string body = resp.Content.ReadAsStringAsync().GetAwaiter().GetResult();

        Assert.Equal(200, (int)resp.StatusCode);
        Assert.Equal("ok-body", body);
        Assert.Equal(VcrOutcome.Ok, vcr.LastKind);
    }
}
